spur/ssh.py

```diff
@@ -80,7 +80,14 @@ def generate_run_command(self, command_args, store_pid,
             commands.append("echo $$")
 
         if cwd is not None:
-            commands.append("cd {0}".format(escape_sh(cwd)))
+            # 1. Change directory before to check if the command exists
+            # (using command || which) because it might be in this
+            # directory:
+            # $ shell.run(["./foo"], cwd="/some/dir")
+            # 2. If we exit when this cd fails, we can't capture the which_return_code.
+            # To minimize the changes, we run cd again just before exec.
+            # Send stderr to /dev/null to capture only the second cd.
+            commands.append("cd {0} 2>/dev/null".format(escape_sh(cwd)))
```

Comment:
Not especially, although setting an environment variable is a bit risky (since it may affect an environment variable in use). Perhaps something along the lines of (leaving out the redirection for brevity):

```sh
{ cd '/some/path' && echo 0; } || { echo $?; exit 1; }
# Other commands
```

It might be nicer to include the command when echoing the result from `cd` and `which` so that the output is slightly more readable and easier to debug (or when the outputs are misaligned, as happens sometimes when the shell produces extra unexpected output) e.g.

```sh
{ cd '/some/path' && echo cd 0; } || { echo cd $?; exit 1; }
{ { command -v cmd || which cmd; } && echo which 0; } || { echo which $?; exit 1; }
```